### PR TITLE
New Addon: `stage-monitor`

### LINKED ISSUE #1680: Transparent Variable and Different Ask Style
**Transparent Variable and different ask style**

I am requesting you guys to create a addon that will change the variable font and style without that orange markings which is annoying sometimes. Also the ask can also change the theme to the choosen style or font.

### LINKED ISSUE #3590: Customizable Variable Watcher Colour
### Is your request related to a problem? Please describe.

The variable watcher's colour doesn't look great and I would like to be able to customize its colour.

### Describe the solution you'd like

A way to customize the colour of the variable watcher's background, value background and text colour. It should support transparency.

### Describe alternatives you've considered

Using laggy text engines or waiting for Scratch Lab to be released.

### Additional context

This would also override the colour of all other stage watchers (such as lists or the username reporter) and the affect stage setting from editor dark mode.

Related to #1680.

### PR DESCRIPTION
**Pull request by Samq64**
Resolves #3590
Resolves #1680 (The transparent values part, custom fonts is a different issue)

### Changes

Adds a new addon that allows customizing the colour and border radius of stage monitors and their values.

### Reason for changes

Some people don't like the orange colour of variable monitors. This addon also allows customizing stage monitors to make them look pretty cool be adding transparency.

### Tests

Tested on Chromium 115.

image

### CODE REVIEW

**Comment by mxmou** [comment]
> This PR also removes monitors from the editor-dark-mode effect stage setting.

They should still be affected by dark mode if this addon is not enabled.

**Comment by kccuber-scratch** [comment]
I think it should also allow changing the border radius of the stage montiros, like to make it look like the early Scratch 3 mockup

**Comment by Samq64** [comment]
> I think it should also allow changing the border radius of the stage montiros, like to make it look like the early Scratch 3 mockup

Would this be close enough?

image

**Comment by Samq64** [comment]
> > This PR also removes monitors from the editor-dark-mode effect stage setting.
> 
> They should still be affected by dark mode if this addon is not enabled.

I guess I could split the monitor css into a different file, but I'm not sure how to make it only run if an add-on is disabled.

**Comment by Samq64** [comment]
I'm marking this as a draft because I found a couple of minor bugs and am going to implement a radius setting and

**Comment by Samq64** [comment]
Oops.

**Comment by mxmou** [comment]
> > > This PR also removes monitors from the editor-dark-mode effect stage setting.
> > 
> > 
> > They should still be affected by dark mode if this addon is not enabled.
> 
> I guess I could split the monitor css into a different file, but I'm not sure how to make it only run if an add-on is disabled.

Using `"if"`, but that isn't necessary here, just make sure this addon is after `editor-dark-mode` in `addons.json`.

**Comment by mxmou** [comment]
For the normal readout, the border radius should probably affect both the monitor itself and the value.

**Comment by Samq64** [comment]
I finally got GitHub Desktop working again! It wouldn't let me login for some reason.

**Comment by Samq64** [comment]
> Using `"if"`, but that isn't necessary here, just make sure this addon is after `editor-dark-mode` in `addons.json`.

Done.



> For the normal readout, the border radius should probably affect both the monitor itself and the value.

I was going to make it two separate settings.

**Comment by lisa-wolfgang** [comment]
> > This PR also removes monitors from the editor-dark-mode effect stage setting.
> 
> They should still be affected by dark mode if this addon is not enabled.

I disagree -- the project player isn't actually part of the editor, and putting project-player-specific settings in editor addons is relatively confusing.

Related to this point:
https://github.com/ScratchAddons/ScratchAddons/issues/3014
https://github.com/ScratchAddons/ScratchAddons/issues/3899#issuecomment-995268806

**Comment by mxmou** [comment]
> > > This PR also removes monitors from the editor-dark-mode effect stage setting.
> > 
> > 
> > They should still be affected by dark mode if this addon is not enabled.
> 
> I disagree -- the project player isn't actually part of the editor, and putting project-player-specific settings in editor addons is relatively confusing.
> 
> Related to this point: #3014 [#3899 (comment)](https://github.com/ScratchAddons/ScratchAddons/issues/3899#issuecomment-995268806)

I think it makes sense for monitors and other elements on the stage to match the color scheme of the editor though. Ideally they would be affected by editor dark mode in the editor and full screen mode and by website dark mode on the project page.

**Comment by Samq64** [comment]
I don't really know the best way to do this or what should effect what so I'll just mark this as a draft for now.

**Comment by WorldLanguages** [comment]
> I think it makes sense for monitors and other elements on the stage to match the color scheme of the editor though. Ideally they would be affected by editor dark mode in the editor and full screen mode and by website dark mode on the project page.

That's not what currently happens, right? Do we think we should open a discussion about this?
We could also discuss whether full screen should be handled by editor dark mode or website dark mode.

**Comment by Samq64** [comment]
> > I think it makes sense for monitors and other elements on the stage to match the color scheme of the editor though. Ideally they would be affected by editor dark mode in the editor and full screen mode and by website dark mode on the project page.
> 
> That's not what currently happens, right?

The colour of monitors is still set with `editor-dark-mode` but and then overwritten by this addon.

**Comment by mxmou** [comment]
> > I think it makes sense for monitors and other elements on the stage to match the color scheme of the editor though. Ideally they would be affected by editor dark mode in the editor and full screen mode and by website dark mode on the project page.
> 
> That's not what currently happens, right? Do we think we should open a discussion about this? We could also discuss whether full screen should be handled by editor dark mode or website dark mode.

I agree about opening a discussion.

**Comment by Samq64** [comment]
Could someone add the pending label?

**Comment by lisa-wolfgang** [comment]
See https://github.com/ScratchAddons/ScratchAddons/discussions/4863 for discussion on the blocking topic.

**Comment by lisa-wolfgang** [comment]
If it is decided to create a theme addon for the project player, it would likely be built off of this addon and renamed.

@Samq64 As the PR author, would you be okay with this?

**Comment by Samq64** [comment]
Sure, I'll just make a new PR in that case.

**Comment by Samq64** [comment]
Will be superseded by #5217.

**Comment by GrahamSH-LLK** [comment]
Is this still under development?

**Comment by Secret-chest** [comment]
Just a heads-up, the monitors normally use their block colour. If you show a motion reporter, they will be blue. Ideally they should be affected by custom block colours.

**Comment by Samq64** [comment]
> Just a heads-up, the monitors normally use their block colour. If you show a motion reporter, they will be blue. Ideally they should be affected by custom block colours.

That should be fixed in `editor-theme3` and then this addon should be moved below it so it overrides its colours.

**Comment by BlueDome77** [comment]
Somewhat Related: [Monitor Styles](https://github.com/TurboWarp/extensions/pull/579)

**Comment by Samq64** [comment]
Let's remove the border radius setting and just the 2 colour settings so it's less confusing to the user if or when we decide we want to upgrade this into a customizable project player colours addon in the future.

**Comment by Secret-chest** [comment]
Why?

File: addons/editor-dark-mode/stage.css [hunk 1]
```diff
@@ -1,18 +1,3 @@
...
-[class*="monitor_list-index_"] {
-  color: var(--editorDarkMode-page-text);
-}
-[class*="monitor_label_"],
-[class*="monitor_list-empty_"] {
-  color: var(--editorDarkMode-page-text) !important;
-}
-
-[class*="monitor_list-header_"],
-[class*="monitor_list-footer_"],
```
Comment (TheColaber) [inline addons/editor-dark-mode/stage.css:15]: i dont think this should be removed. dark mode should still apply changes unless this addon is enabled, where in that case, this addon should override darkmode.

Comment (Samq64) [reply addons/editor-dark-mode/stage.css:15]: I guess that makes sense since the setting is still there for other things.

**Review by TheColaber — APPROVED** [review]

File: addons/stage-monitor/addon.json [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      "id": "customValueColor",
+      "type": "boolean",
+      "default": false
+    },
+    {
+      "name": "Value color",
+      "id": "monitorValueBg",
+      "type": "color",
+      "allowTransparency": true,
+      "default": "#333333",
```
Comment (mxmou) [inline addons/stage-monitor/addon.json:101]:
```suggestion
      "default": "#575e75",
```
The default should be a color from Scrath's color palette.

File: addons/stage-monitor/addon.json [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      "if": {
+        "settings": { "customValueColor": true }
+      }
+    }
+  ],
+  "presets": [
+    {
+      "name": "Transparent values",
+      "id": "transparent",
+      "description": "Scratch's default monitor colors with transparent value backgrounds.",
```
Comment (mxmou) [inline addons/stage-monitor/addon.json:111]:
```suggestion
      "description": "Scratch's default monitor colors without value backgrounds.",
```

File: addons/stage-monitor/addon.json [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    },
+    {
+      "name": "Translucent black",
+      "id": "translucent",
+      "description": "Translucent black monitors like in HTMLifier.",
+      "values": {
+        "monitor": "#00000040",
+        "listHeader": "#00000040",
+        "monitorRadius": 4,
+        "valueRadius": 4,
```
Comment (mxmou) [inline addons/stage-monitor/addon.json:140]: Unless a preset that changes them is added, I don't think presets should set these.

**Review by mxmou — APPROVED** [review]

File: addons/stage-monitor/addon.json [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        "monitorValueBg": "#00000040"
+      }
+    }
+  ],
+  "dynamicEnable": true,
+  "dynamicDisable": true,
+  "updateUserstylesOnSettingsChange": true,
+  "injectAsStyleElt": true,
+  "tags": ["editor", "theme", "projectPlayer"],
+  "versionAdded": "1.40.0"
```
Comment (Samq64) [inline addons/stage-monitor/addon.json:146]: Just a reminder.

Comment (Samq64) [reply addons/stage-monitor/addon.json:146]: I meant a reminder to update this, not that we need to wait to merge it.

**Comment by WorldLanguages** [comment]
We can do v1.39 for this one, strings aren't fully closed yet

**Comment by Samq64** [comment]
I am now 15% older.

**Pull request merged**